Detour/Source/DetourNavMeshQuery.cpp

```diff
@@ -98,8 +98,114 @@ inline float dtQueryFilter::getCost(const float* pa, const float* pb,
 {
 	return dtVdist(pa, pb) * m_areaCost[curPoly->getArea()];
 }
-#endif	
-	
+#endif
+
+
+dtQueryPathState* dtAllocQueryPathState()
+{
+	void* mem = dtAlloc(sizeof(dtQueryPathState), DT_ALLOC_PERM);
+	if (!mem) return 0;
+	return new(mem) dtQueryPathState;
+}
+
+void dtFreeQueryPathState(dtQueryPathState* state)
+{
+	if (!state) return;
+	state->~dtQueryPathState();
+	dtFree(state);
+}
+
+dtQueryPathState::dtQueryPathState() :
+	m_nodePool(0),
+	m_openList(0),
+	m_startRefs(0),
+	m_numStarts(0)
+{};
+
+
+dtQueryPathState::~dtQueryPathState()
+{
+	if (m_nodePool)
+		m_nodePool->~dtNodePool();
+	if (m_openList)
+		m_openList->~dtNodeQueue();
+
+
+	dtFree(m_nodePool);
+	dtFree(m_openList);
+	dtFree(m_startRefs);
+}
+
+dtStatus dtQueryPathState::init(const int maxNodes)
+{
+	if (maxNodes > DT_NULL_IDX || maxNodes > (1 << DT_NODE_PARENT_BITS) - 1)
+		return DT_FAILURE | DT_INVALID_PARAM;
+
+	if (!m_nodePool || m_nodePool->getMaxNodes() < maxNodes)
+	{
+		if (m_nodePool)
+		{
+			m_nodePool->~dtNodePool();
```

Comment:
why not `m_nodePool->clear();` ?

Reply:
This is calling the destructor. clear() just clears the internal data structures, it doesn’t deallocate them.